sdk/formrecognizer/ai-form-recognizer/test/node/setup.spec.ts

```diff
@@ -0,0 +1,109 @@
+ */
+
+import fs from "fs-extra";
+import path from "path";
+
+import {
+  BlobServiceClient,
+  AnonymousCredential,
+  generateBlobSASQueryParameters,
+  StorageSharedKeyCredential,
+  ContainerSASPermissions,
+  SASProtocol
+} from "@azure/storage-blob";
+import * as recorder from "@azure/test-utils-recorder";
+
+import { setTrainingContainerSasUrl } from "../util/trainingContainer";
+
+import * as dotenv from "dotenv";
+dotenv.config();
+
+if (!recorder.isPlaybackMode()) {
+  const ASSET_PATH = path.join(__dirname, "..", "test-assets", "training");
+
+  const blobEndpoint = process.env.BLOB_SAS_ENDPOINT;
+  if (blobEndpoint === undefined) {
+    throw new Error("No SAS URL provided for training data account");
+  }
+
+  const trainingContainerName = `training${new Date().getTime()}`;
+
+  const [blobAccountName, blobAccountKey] = [
+    process.env.BLOB_ACCOUNT_NAME,
+    process.env.BLOB_ACCESS_KEY
+  ];
+
+  if (blobAccountName === undefined || blobAccountKey === undefined) {
+    throw new Error(
+      "Both BLOB_ACCOUNT_NAME and BLOB_ACCOUNT_KEY must be provided in the environment."
+    );
+  }
+
+  const blobClient = new BlobServiceClient(blobEndpoint, new AnonymousCredential());
+
+  // Set up training container
+  before(async () => {
+    console.log("=== Running pre-test hook to upload training data ===");
+
+    const { containerClient: trainingContainer } = await blobClient.createContainer(
+      trainingContainerName
+    );
+
+    const directories = [ASSET_PATH];
+    while (directories.length > 0) {
+      const directory = directories.shift() as string;
+      // Upload all files in training folder to the container
+      for (let baseName of await fs.readdir(directory)) {
+        const fileName = path.join(directory, baseName);
+
+        const stats = await fs.stat(fileName);
+
+        if (stats.isDirectory()) {
+          directories.push(fileName);
+        } else {
+          const blobName = path.relative(ASSET_PATH, fileName);
+          console.log(`- Uploading ${blobName} ...`);
+
+          await trainingContainer.uploadBlockBlob(
+            blobName,
+            () => fs.createReadStream(fileName),
+            stats.size
+          );
+        }
+      }
+    }
+
+    console.log("Done.");
+
+    const oneHourFromNow = new Date(new Date().valueOf() + 60 * 60 * 1000);
+
+    const sasQueryParameters = generateBlobSASQueryParameters(
+      {
+        containerName: trainingContainer.containerName,
+        permissions: ContainerSASPermissions.parse("rl"),
+        startsOn: new Date(),
+        expiresOn: oneHourFromNow, // Tests time out in an hour any way
+        protocol: SASProtocol.Https
+      },
+      new StorageSharedKeyCredential(blobAccountName, blobAccountKey)
+    );
+
+    const sasUrl = blobEndpoint.split("?")[0] + "?" + sasQueryParameters.toString();
+
+    setTrainingContainerSasUrl(sasUrl);
+  });
+
+  // Tear down training container
+  after(async () => {
```

Comment:
should we rename this test to `training.spec.ts` and add training tests here? otherwise the container and training data would be deleted.

Reply:
I think an `after()` hook in the global scope runs after all of the test suites, so this `before()` will run first thing, before any tests, and the `after()` will run at the very end. We certainly could put all of our training tests here and encapsulate them in the same place, but I figured it would be slightly simpler to setup/teardown at the very beginning and end.

For the tests, we will probably have to train a model in the setup for other test units to use, so we will have to be able to know that the training takes place before any of those units run.

Reply:
I see. This is the global level hooks outside of any describe() block.

We can put training and custom form recognition into one tests because we need to know the trained model Ids and it's easier in one test than passing between tests.